core/executor/src/allocator.rs

```diff
@@ -155,22 +156,26 @@ impl FreeingBumpHeapAllocator {
 
 	fn get_heap_4bytes(&mut self, ptr: u32) -> [u8; 4] {
 		let mut arr = [0u8; 4];
-		self.heap.get_into(self.ptr_offset + ptr, &mut arr).unwrap();
+		self.heap.get_into(self.ptr_offset + ptr, &mut arr)
+			.expect("Getting 4 bytes from heap at pointer must always be valid");
```

Comment:
But what guarantees that?

Reply:
To the best of our knowledge the calculation of the pointer with which this function is called is correct and there are a number of tests which test for that. 

Do you suggest to amend the comment itself to sth like `Getting 4 bytes from heap at pointer is always valid because this function is only called with valid addresses` or to introduce some validation logic (which explicitly tests if the pointer is valid) into this function?

Reply:
it is better to localize the assumptions

Reply:
But the real reason you can't write a valid `expect` here is that the code is broken :)

https://github.com/paritytech/substrate/blob/072b396ce52d936ccbe3e193bcb1bcaef0dc2ca3/core/executor/src/allocator.rs#L121

And actually as far as I can see the `deallocate` function (nor its call sites: `ext_free` or `sandbox::deallocate`) is not 
  1. checking that `ptr > 8` -- integer underflow is possible
  2. ensuring that the `ptr` is actually allocated by the allocator.

Since requirement 2 is probably not efficient to check, it's better just to handle the case where there is no data in the heap at that pointer.

Sure, runtimes aren't supposed to call `deallocate` with invalid data, and mostly they won't, but _we are not going to panic if they do_.

Reply:
Yup - if you deallocate and the previous 8-bytes is not `0x__ffffffffffffff`, then you've got heap corruption; similarly if you allocate and the previous 8-bytes is an invalid offset into memory, then again you've got heap corruption. Non-`unsafe` Rust should prevent this from happening ever, but we cannot guarantee that the runtime will be Rust-compiled, or that it didn't get corrupted on its way into the chain. In all instances, we should report heap corruption and act in the same way as if memory was exhausted (i.e. end execution immediately with an appropriate `Error`).

Reply:
Ok, I have amended the error handling so that we no longer panic, but instead report an error which results in a trap. Is this stopping execution immediate enough?

I'm not entirely happy with returning `UserError` , but this type is used for all kinds of errors regarding the sandbox.